pkg/operator/minio/controller.go

```diff
@@ -0,0 +1,299 @@
+	coreV1Client := c.context.Clientset.CoreV1()
+
+	svc, err := coreV1Client.Services(namespace).Create(&v1.Service{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:            name,
+			Namespace:       namespace,
+			Labels:          map[string]string{k8sutil.AppAttr: minioLabel},
+			OwnerReferences: []meta_v1.OwnerReference{ownerRef},
+		},
+		Spec: v1.ServiceSpec{
+			Selector:  map[string]string{k8sutil.AppAttr: minioLabel},
+			Ports:     []v1.ServicePort{{Port: spec.Port}},
+			ClusterIP: v1.ClusterIPNone,
+		},
+	})
+
+	return svc, err
+}
+
+func (c *MinioController) buildMinioCtrArgs(statefulSetPrefix, headlessServiceName, namespace string, serverCount int32) []string {
+	args := []string{"server"}
+	for i := int32(0); i < serverCount; i++ {
+		serverAddress := fmt.Sprintf("http://%s-%d.%s.%s%s", statefulSetPrefix, i, headlessServiceName, fmt.Sprintf(minioServerSuffixFmt, namespace), objectStoreDataDir)
+		args = append(args, serverAddress)
+	}
+
+	logger.Infof("Building Minio container args: %v", args)
+	return args
+}
+
+func (c *MinioController) makeMinioPodSpec(name, namespace string, ctrName string, ctrImage string, port int32, envVars map[string]string, numServers int32) v1.PodTemplateSpec {
+	var env []v1.EnvVar
+	for k, v := range envVars {
+		env = append(env, v1.EnvVar{Name: k, Value: v})
+	}
+
+	podSpec := v1.PodTemplateSpec{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    map[string]string{k8sutil.AppAttr: minioLabel},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    ctrName,
+					Image:   ctrImage,
+					Env:     env,
+					Command: []string{"/usr/bin/minio"},
+					Ports:   []v1.ContainerPort{{ContainerPort: port}},
+					Args:    c.buildMinioCtrArgs(name, name, namespace, numServers),
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      minioVolumeName,
+							MountPath: objectStoreDataDir,
+						},
+					},
+				},
+			},
+			Volumes: []v1.Volume{
+				{
+					Name: minioVolumeName,
+					VolumeSource: v1.VolumeSource{
+						PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+							ClaimName: minioPVCName,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return podSpec
+}
+
+func (c *MinioController) getAccessCredentials(secretName, namespace string) (string, string, error) {
+	coreV1Client := c.context.Clientset.CoreV1()
+	var getOpts meta_v1.GetOptions
+	val, err := coreV1Client.Secrets(namespace).Get(secretName, getOpts)
+	if err != nil {
+		logger.Errorf("Unable to get secret with name=%s in namespace=%s: %v", secretName, namespace, err)
+		return "", "", err
+	}
+
+	return string(val.Data["username"]), string(val.Data["password"]), nil
+}
+
+func validateObjectStoreSpec(spec miniov1alpha1.ObjectStoreSpec) error {
+	// Verify node count.
+	count := spec.Storage.NodeCount
+	if count < 4 || count%2 != 0 {
```

Comment:
The simple test scenario is with minikube. Are you saying it's required to start a 4-node cluster?

Reply:
The terminology here is that Node means an instance of the distributed storage system, while machine/server is the physical concept.  It is perfectly possible to have more than 1 node on a single machine/server.  Take for example the [CockroachDB doc](https://www.cockroachlabs.com/docs/stable/start-a-local-cluster.html#step-2-add-nodes-to-the-cluster):
> Since you're running all nodes on the same machine, you also set the --store, --port, and --http-port flags to locations and ports not used by other nodes, but in a real deployment, with each node on a different machine, the defaults would suffice.